docs/src/adr/014adr-precise-records.md

```diff
@@ -0,0 +1,668 @@
+Unfortunately, variants introduce additional complexity.
+
+### 5.2. Variants
+
+**Example 5.2.1.** Now we have to figure out how to deal with TLA+ expressions
+like:
+
+```tla
+  { [ tag |-> "1a", bal |-> 3 ], [ tag |-> "2a", bal |-> 4, val |-> 0 ] }
+```
+
+Obviously, we cannot fit both of the records into a single plain record type,
+provided that we want to precisely track the fields that are present in a
+record.  So the type checker should report a type error, if we only implement
+type inference for the case explained in Section 5.1. To support this important
+pattern, we introduce variants. They are similar to [unions in
+TypeScript](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#union-types).
+In contrast to TypeScript, we fix one field to designate the record type in a
+variant. Also, we are using the word "variant", to avoid any confusion
+with the TLA+ operators `UNION` and `\union`.
+
+We reserve the field name `tag` for the record discriminator.
+
+**Variant constructor.** We introduce a special TLA+ operator `Variant`
+that extracts the tag from a record and wraps the record into a variant.
+We need this operator to distinguish between plain records and records that
+belong to a variant. The operator `Variant` is defined in TLA+ as follows:
+
+```tla
+  Variant(r) ==
+    \* fallback untyped implementation
+    r
+```
+
+This operator does not change its argument, but it provides the type checker
+with a hint that the record should be treated as a member of a variant.
+
+Consider the record constructor of `n+1` fields, one of them being the field
+`"tag"`:
+
+```tla
+  [ tag |-> "<TAG>", f_1 |-> e_1, ..., f_n |-> e_n ]
+```
+
+Here, `"<TAG>"` stands for a string literal such as `"1a"` or `"2a"`. The
+general rule for `Variant([ tag |-> "<TAG>", f_1 |-> e_1, ..., f_n |-> e_n ])`
+looks as follows:
+
+```
+e_1: t_1, ..., e_n: t_n
+z is a fresh type variable
+------------------------------------------------------------ [variant]
+Variant([ tag |-> "<TAG>", f_1 |-> e_1, ..., f_n |-> e_n ]):
+  Variant(
+    RowCons("<TAG>",
+      Rec(
+        RowCons("tag", Str,
+          RowCons(f_1, t_1,
+            ...
+              RowNil)...)
+      ),
+      z
+    )
+  )
+```
+
+According to the rule `[variant]`, the operator `Variant` wraps a record
+constructor that contains a string literal for the field `tag`. The variant
+contains the record that was passed in the constructor, whereas the other
+elements of the variant are captured with a fresh type variable `z`, which must
+be a row.
+
+Importantly, we use rows at two levels:
+
+1. To construct a single record, whose shape is defined precisely.
+
+1. To construct a variant, whose only record is known at the time,
+   while the rest is captured with the type variable `z`.
+
+Going back to **Example 5.2.1**, the set constructor would produce a set of
+equalities:
+
+```
+  a = Set(b)
+  a = Set(d)
```

Comment:
What is the reason for using `a` twice here?

Reply:
It's just the form of the constructed constraints. The left-hand side is alway a type variable. We could write it as one equality.